library/psa_crypto.c

```diff
@@ -3887,7 +3874,707 @@ psa_status_t psa_aead_decrypt( mbedtls_svc_key_id_t key,
+                                                          operation->nonce_length,
+                                                          input,
+                                                          input_length ) );
+
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_GCM */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)
+    if( operation->alg == PSA_ALG_CCM )
+    {
+        /* Multipart CCM not supported as yet */
+        (void) input;
+        (void) input_length;
+
+        return ( PSA_ERROR_NOT_SUPPORTED );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_CCM */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305)
+    if( operation->alg == PSA_ALG_CHACHA20_POLY1305 )
+    {
+        status = mbedtls_to_psa_error( mbedtls_chachapoly_update_aad( &operation->ctx.chachapoly,
+                                                                      input,
+                                                                      input_length ) );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305 */
+    {
+        (void) input;
+        (void) input_length;
+
+        return ( PSA_ERROR_NOT_SUPPORTED );
+    }
+
+    if( status == PSA_SUCCESS )
+    {
+        operation->ad_started = 1;
+    }
 
+    return ( status );
+}
+
+/* Encrypt or decrypt a message fragment in an active multipart AEAD
+ * operation.*/
+psa_status_t psa_aead_update(psa_aead_operation_t *operation,
+                             const uint8_t *input,
+                             size_t input_length,
+                             uint8_t *output,
+                             size_t output_size,
+                             size_t *output_length)
+{
+    size_t update_output_size;
+    size_t buffer_fill_length = 0;
+    size_t input_length_left = input_length;
+    size_t input_block_length = 0;
+    psa_status_t status;
+
+    if( !operation->nonce_set || !operation->key_set || !operation->ad_started )
+    {
+        return( PSA_ERROR_BAD_STATE );
+    }
+
+    update_output_size = PSA_AEAD_UPDATE_OUTPUT_SIZE(operation->alg, input_length);
+
+    if(update_output_size > output_size )
+    {
+        return ( PSA_ERROR_BUFFER_TOO_SMALL );
+    }
+
+    if( operation->lengths_set)
+    {
+        /* Additional data length was supplied, but not all the additional
+           data was supplied.*/
+        if( operation->ad_remaining != 0 )
+        {
+            return ( PSA_ERROR_INVALID_ARGUMENT );
+        }
+
+        /* Too much data provided. */
+        if( operation->body_remaining < input_length )
+        {
+            return ( PSA_ERROR_INVALID_ARGUMENT );
+        }
+
+        operation->body_remaining -= input_length;
+    }
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)
+    if( operation->alg == PSA_ALG_GCM )
+    {
+        if( !operation->lengths_set )
+        {
+            return( PSA_ERROR_BAD_STATE );
+        }
+
+        /* GCM should only be presented with block size chunks, so ensure
```

Comment:
It seems to me that there's more code here to work around this limitation of `mbedtls_gcm_update` than it would take to modify `mbedtls_gcm_update` to lift this limitation. And at first glance we can do this without modifying the layout of `mbedtls_gcm_context`.

Reply:
Done in https://github.com/ARMmbed/mbedtls/pull/4342